File: airbyte-server/src/test/java/io/airbyte/server/apis/UserApiControllerTest.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis;

import io.airbyte.api.model.generated.OrganizationIdRequestBody;
import io.airbyte.api.model.generated.OrganizationUserReadList;
import io.airbyte.api.model.generated.UserAuthIdRequestBody;
import io.airbyte.api.model.generated.UserCreate;
import io.airbyte.api.model.generated.UserEmailRequestBody;
import io.airbyte.api.model.generated.UserGetOrCreateByAuthIdResponse;
import io.airbyte.api.model.generated.UserIdRequestBody;
import io.airbyte.api.model.generated.UserRead;
import io.airbyte.api.model.generated.UserUpdate;
import io.airbyte.api.model.generated.UserWithPermissionInfoReadList;
import io.airbyte.api.model.generated.WorkspaceIdRequestBody;
import io.airbyte.api.model.generated.WorkspaceUserReadList;
import io.airbyte.commons.json.Jsons;
import io.airbyte.config.persistence.ConfigNotFoundException;
import io.airbyte.validation.json.JsonValidationException;
import io.micronaut.context.annotation.Requires;
import io.micronaut.context.env.Environment;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpStatus;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import java.io.IOException;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;

@MicronautTest
@Requires(env = {Environment.TEST})
@SuppressWarnings("PMD.JUnitTestsShouldIncludeAssert")
class UserApiControllerTest extends BaseControllerTest {

  @Test
  void testCreateUser() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.when(userHandler.createUser(Mockito.any()))
        .thenReturn(new UserRead());
    final String path = "/api/v1/users/create";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserCreate())),
        HttpStatus.OK);
  }

  @Test
  void testGetUser() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.when(userHandler.getUser(Mockito.any()))
        .thenReturn(new UserRead());
    final String path = "/api/v1/users/get";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserIdRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testGetUserByAuthId() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.when(userHandler.getUserByAuthId(Mockito.any()))
        .thenReturn(new UserRead());
    final String path = "/api/v1/users/get_by_auth_id";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserAuthIdRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testGetUserByEmail() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.when(userHandler.getUserByEmail(Mockito.any()))
        .thenReturn(new UserRead());
    final String path = "/api/v1/users/get_by_email";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserEmailRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testDeleteUser() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.doNothing().when(userHandler).deleteUser(Mockito.any());
    final String path = "/api/v1/users/delete";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserIdRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testUpdateUser() throws JsonValidationException, ConfigNotFoundException, IOException {
    Mockito.when(userHandler.updateUser(Mockito.any()))
        .thenReturn(new UserRead());
    final String path = "/api/v1/users/update";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserUpdate())),
        HttpStatus.OK);
  }

  @Test
  void testListUsersInOrganization() throws IOException, ConfigNotFoundException {
    Mockito.when(userHandler.listUsersInOrganization(Mockito.any()))
        .thenReturn(new OrganizationUserReadList());
    final String path = "/api/v1/users/list_by_organization_id";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new OrganizationIdRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testListUsersInWorkspace() throws Exception {
    Mockito.when(userHandler.listUsersInWorkspace(Mockito.any()))
        .thenReturn(new WorkspaceUserReadList());
    final String path = "/api/v1/users/list_by_workspace_id";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new WorkspaceIdRequestBody())),
        HttpStatus.OK);
  }

  @Test
  void testListInstanceAdminUsers() throws Exception {
    Mockito.when(userHandler.listInstanceAdminUsers())
        .thenReturn(new UserWithPermissionInfoReadList());
    final String path = "/api/v1/users/list_instance_admin";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.emptyObject()),
        HttpStatus.OK);
  }

  @Test
  void testGetOrCreateUser() throws Exception {
    Mockito.when(userHandler.getOrCreateUserByAuthId(Mockito.any()))
        .thenReturn(new UserGetOrCreateByAuthIdResponse().userRead(new UserRead()));
    final String path = "/api/v1/users/get_or_create_by_auth_id";
    testEndpointStatus(
        HttpRequest.POST(path, Jsons.serialize(new UserAuthIdRequestBody())),
        HttpStatus.OK);
  }

}

File: airbyte-server/src/test/kotlin/io/airbyte/server/apis/controllers/UserApiControllerTest.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.apis.controllers

import io.airbyte.api.model.generated.OrganizationIdRequestBody
import io.airbyte.api.model.generated.OrganizationUserReadList
import io.airbyte.api.model.generated.UserAuthIdRequestBody
import io.airbyte.api.model.generated.UserEmailRequestBody
import io.airbyte.api.model.generated.UserGetOrCreateByAuthIdResponse
import io.airbyte.api.model.generated.UserIdRequestBody
import io.airbyte.api.model.generated.UserRead
import io.airbyte.api.model.generated.UserUpdate
import io.airbyte.api.model.generated.UserWithPermissionInfoReadList
import io.airbyte.api.model.generated.WorkspaceIdRequestBody
import io.airbyte.api.model.generated.WorkspaceUserAccessInfoReadList
import io.airbyte.commons.json.Jsons
import io.airbyte.commons.server.handlers.UserHandler
import io.airbyte.server.assertStatus
import io.airbyte.server.status
import io.micronaut.http.HttpRequest
import io.micronaut.http.HttpStatus
import io.micronaut.http.client.HttpClient
import io.micronaut.http.client.annotation.Client
import io.micronaut.test.annotation.MockBean
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import io.mockk.every
import io.mockk.mockk
import jakarta.inject.Inject
import org.junit.jupiter.api.Test

@MicronautTest
internal class UserApiControllerTest {
  @Inject
  lateinit var userHandler: UserHandler

  @Inject
  @Client("/")
  lateinit var client: HttpClient

  @MockBean(UserHandler::class)
  fun userHandler(): UserHandler = mockk()

  @Test
  fun testGetUser() {
    every { userHandler.getUser(any()) } returns UserRead()

    val path = "/api/v1/users/get"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserIdRequestBody())))
  }

  @Test
  fun testGetUserByAuthId() {
    every { userHandler.getUserByAuthId(any()) } returns UserRead()

    val path = "/api/v1/users/get_by_auth_id"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserAuthIdRequestBody())))
  }

  @Test
  fun testGetUserByEmail() {
    every { userHandler.getUserByEmail(any()) } returns UserRead()

    val path = "/api/v1/users/get_by_email"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserEmailRequestBody())))
  }

  @Test
  fun testDeleteUser() {
    every { userHandler.deleteUser(any()) } returns Unit

    val path = "/api/v1/users/delete"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserIdRequestBody())))
  }

  @Test
  fun testUpdateUser() {
    every { userHandler.updateUser(any()) } returns UserRead()

    val path = "/api/v1/users/update"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserUpdate())))
  }

  @Test
  fun testListUsersInOrganization() {
    every { userHandler.listUsersInOrganization(any()) } returns OrganizationUserReadList()

    val path = "/api/v1/users/list_by_organization_id"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, OrganizationIdRequestBody())))
  }

  @Test
  fun testListInstanceAdminUsers() {
    every { userHandler.listInstanceAdminUsers() } returns UserWithPermissionInfoReadList()

    val path = "/api/v1/users/list_instance_admin"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, Jsons.emptyObject())))
  }

  @Test
  fun testGetOrCreateUser() {
    every { userHandler.getOrCreateUserByAuthId(any()) } returns UserGetOrCreateByAuthIdResponse().userRead(UserRead())

    val path = "/api/v1/users/get_or_create_by_auth_id"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, UserAuthIdRequestBody())))
  }

  @Test
  fun testListAccessInfoByWorkspaceId() {
    every { userHandler.listAccessInfoByWorkspaceId(any()) } returns WorkspaceUserAccessInfoReadList()

    val path = "/api/v1/users/list_access_info_by_workspace_id"
    assertStatus(HttpStatus.OK, client.status(HttpRequest.POST(path, Jsons.serialize(WorkspaceIdRequestBody()))))
  }
}
